core/src/main/java/com/backblaze/b2/client/B2LargeFileStorer.java

```diff
@@ -0,0 +1,209 @@
+                                                           List<Future<B2Part>> partFutures) throws B2Exception {
+
+        final List<String> partSha1s = new ArrayList<>();
+        try {
+            for (final Future<B2Part> partFuture : partFutures) {
+                final B2Part part = partFuture.get();
+                partSha1s.add(part.getContentSha1());
+            }
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            throw new B2LocalException("interrupted", "interrupted while trying to copy parts: " + e, e);
+
+        } catch (ExecutionException e) {
+            Throwable cause = e.getCause();
+            if (cause instanceof B2Exception) {
+                throw (B2Exception) e.getCause();
+            } else {
+                throw new B2LocalException("trouble", "exception while trying to upload parts: " + cause, cause);
+            }
+        } finally {
+            // we've either called get() on all of the futures, or we've hit an exception and
+            // we aren't going to wait for the others.  let's call cancel on all of them.
+            // the ones that have finished already, won't mind and the others will be stopped.
+            for (Future<B2Part> future : partFutures) {
+                future.cancel(true);
+            }
+        }
+
+        // finish the large file.
+        B2FinishLargeFileRequest finishRequest = B2FinishLargeFileRequest
+                .builder(largeFileVersion.getFileId(), partSha1s)
+                .build();
+        return retryer.doRetry(
+                "b2_finish_large_file",
+                accountAuthCache,
+                () -> webifier.finishLargeFile(accountAuthCache.get(), finishRequest),
+                retryPolicySupplier.get());
+    }
+
+    B2Part uploadPart(int partNumber, B2ContentSource contentSource) throws B2Exception {
+
+        return retryer.doRetry(
+                "b2_upload_part",
+                accountAuthCache,
+                (isRetry) -> {
+                    final B2UploadPartUrlResponse uploadPartUrlResponse = uploadPartUrlCache.get(isRetry);
+
+                    final B2UploadPartRequest uploadPartRequest = B2UploadPartRequest
+                            .builder(partNumber, contentSource)
+                            .build();
+
+                    final B2Part part = webifier.uploadPart(uploadPartUrlResponse, uploadPartRequest);
+
+                    // Return the upload part URL, because it works and can be reused.
+                    uploadPartUrlCache.unget(uploadPartUrlResponse);
+
+                    return part;
+                },
+                retryPolicySupplier.get()
+        );
+    }
+
+    B2Part copyPart(int partNumber, String sourceFileId, B2ByteRange byteRangeOrNull) throws B2Exception {
+        final B2CopyPartRequest copyPartRequest = B2CopyPartRequest
+                .builder(sourceFileId, fileVersion.getFileId(), partNumber)
```

Comment:
in the category of "observation i'm not sure what to ask for": slightly interesting that the parameters to builder() are in a different order than the ones for copyPart().

Reply:
I went ahead and put `partNumber` first.